mindmeld/components/entity_resolver.py

```diff
@@ -42,60 +50,277 @@
+        return "double_metaphone" in self.er_config.get("model_settings", {}).get(
+            "phonetic_match_types", [])
+
+    @staticmethod
+    def _process_entity_map(entity_type, entity_map, normalizer=None, augment_lower_case=False,
+                            augment_title_case=False, augment_normalized=False):
+        """
+        Loads in the mapping.json file and stores the synonym mappings in a item_map
+            and a synonym_map
+
+        Args:
+            entity_type (str): The entity type associated with this entity resolver
+            entity_map (dict): The loaded mapping.json file for the given entity type
+            normalizer (callable): The normalizer to use, if provided, used to normalize synonyms
+            augment_lower_case (bool): If to extend the synonyms list with their lower-cased values
+            augment_title_case (bool): If to extend the synonyms list with their title-cased values
+            augment_normalized (bool): If to extend the synonyms list with their normalized values,
+                                        uses the provided normalizer
+        """
+
+        if augment_lower_case or augment_title_case or augment_normalized:
+            def _form_explorer(string):
+                _forms = []
+                if augment_lower_case:
+                    _forms.append(string.lower())
+                if augment_title_case:
+                    _forms.append(string.title())
+                if augment_normalized and normalizer:
+                    _forms.append(normalizer(string))
+                return _forms
+        else:
+            _form_explorer = None
+
+        item_map = {}
+        syn_map = {}
+        seen_ids = []
+        for item in entity_map.get("entities", []):
+            cname = item["cname"]
+            item_id = item.get("id")
+            if cname in item_map:
+                msg = "Canonical name %s specified in %s entity map multiple times"
+                logger.debug(msg, cname, entity_type)
+            if item_id:
+                if item_id in seen_ids:
+                    msg = "Item id {!r} specified in {!r} entity map multiple times"
+                    raise ValueError(msg.format(item_id, entity_type))
+                seen_ids.append(item_id)
+
+            aliases = [cname] + item.pop("whitelist", [])
+            items_for_cname = item_map.get(cname, [])
+            items_for_cname.append(item)
+            item_map[cname] = items_for_cname
+            for alias in aliases:
+                norm_alias = normalizer(alias) if normalizer else alias
+                if norm_alias in syn_map:
+                    msg = "Synonym %s specified in %s entity map multiple times"
+                    logger.debug(msg, cname, entity_type)
+                cnames_for_syn = syn_map.get(norm_alias, [])
+                cnames_for_syn.append(cname)
+                syn_map[norm_alias] = list(set(cnames_for_syn))
```

Comment:
I think that the `if _form_explorer:` logic can be integrated into this loop directly. We have to augment the `list(set(cnames_for_syn))` with the relevant mutations.